Sources/NIOSSL/SecurityFrameworkCertificateVerification.swift

```diff
@@ -47,11 +47,40 @@ extension SSLConnection {
             var trust: SecTrust? = nil
             var result: OSStatus
             let policy = SecPolicyCreateSSL(self.role! == .client, self.expectedHostname as CFString?)
-            result = SecTrustCreateWithCertificates(certificates as CFArray, policy, &trust)
+            result = SecTrustCreateWithCertificates(peerCertificates as CFArray, policy, &trust)
             guard result == errSecSuccess, let actualTrust = trust else {
                 throw NIOSSLError.unableToValidateCertificate
             }
 
+            // If there are additional trust roots then we need to add them to the SecTrust as anchors.
+            let additionalAnchorCertificates: [SecCertificate] = try self.parentContext.configuration.additionalTrustRoots.flatMap { trustRoots -> [NIOSSLCertificate] in
+                switch trustRoots {
+                case .file(let path):
+                    return try NIOSSLCertificate.fromPEMFile(path)
+                case .certificates(let certs):
+                    return certs
+                }
+            }.map {
+                guard let secCert = SecCertificateCreateWithData(nil, Data(try $0.toDERBytes()) as CFData) else {
+                    throw NIOSSLError.failedToLoadCertificate
+                }
+                return secCert
+            }
+            if !additionalAnchorCertificates.isEmpty {
+                // To use additional anchors _and_ the built-in ones we must reenable the built-in ones expicitly.
+                switch self.parentContext.configuration.trustRoots {
+                case .none, .some(.default):
+                    guard SecTrustSetAnchorCertificatesOnly(actualTrust, false) == errSecSuccess else {
+                        throw NIOSSLError.failedToLoadCertificate
+                    }
+                default:
+                    break
```

Comment:
I don't think `break` is right: I think we can only get here in the above branch case. We can again just `preconditionFailure` here.

Reply:
Again this was done for loose coupling reasons. E.g. if we wanted to make use of Security.framework for the case where we have trust roots and additional trust roots but none of them are the platform default (e.g. some files and some `NIOSSLCertificate` values).

I know that currently in this case the code currently goes through BoringSSL for that configuration but do we want to bake that into this layer too?

Reply:
I think we do. Specifically, I think the case where we change the behaviour of the verification code will make us want to think harder about what we're doing, and I'd rather that instead of silently appearing to work it loudly failed. Preconditions can always be removed.

Reply:
Addressed in https://github.com/apple/swift-nio-ssl/pull/284/commits/9d613d303a6623c384855ea67b52af5a318aeea6.